libraries/AP_Compass/AP_Compass_BMM350.cpp

```diff
@@ -0,0 +1,492 @@
+}
+
+void AP_Compass_BMM350::timer()
+{
+    struct PACKED {
+        uint8_t magx[3];
+        uint8_t magy[3];
+        uint8_t magz[3];
+        uint8_t temp[3];
+    } data;
+
+    // Read out measurement data
+    if (!read_bytes(BMM350_REG_MAG_X_XLSB, (uint8_t *)&data, sizeof(data))) {
+        return;
+    }
+
+    // Converts 24-bit raw data to signed integer value
+    const int32_t magx_raw = int32_t(((uint32_t)data.magx[0] << 8) + ((uint32_t)data.magx[1] << 16) + ((uint32_t)data.magx[2] << 24)) >> 8;
+    const int32_t magy_raw = int32_t(((uint32_t)data.magy[0] << 8) + ((uint32_t)data.magy[1] << 16) + ((uint32_t)data.magy[2] << 24)) >> 8;
+    const int32_t magz_raw = int32_t(((uint32_t)data.magz[0] << 8) + ((uint32_t)data.magz[1] << 16) + ((uint32_t)data.magz[2] << 24)) >> 8;
+    const int32_t temp_raw = int32_t(((uint32_t)data.temp[0] << 8) + ((uint32_t)data.temp[1] << 16) + ((uint32_t)data.temp[2] << 24)) >> 8;
+    
+    // Convert mag lsb to uT and temp lsb to degC
+    float magx = (float)magx_raw * BMM350_XY_SCALE;
+    float magy = (float)magy_raw * BMM350_XY_SCALE;
+    float magz = (float)magz_raw * BMM350_Z_SCALE;
+    float temp = (float)temp_raw * BMM350_TEMP_SCALE;
+
+    if (temp > 0.0f) {
+        temp -= 25.49f;
+    } else if (temp < 0.0f) {
+        temp += 25.49f;
+    }
+
+    // Apply compensation
+    temp = ((1 + _mag_comp.sensit_coef.temp) * temp) + _mag_comp.offset_coef.temp;
+    // Compensate raw magnetic data
+    magx = ((1 + _mag_comp.sensit_coef.x) * magx) + _mag_comp.offset_coef.x + (_mag_comp.tco.x * (temp - _mag_comp.t0_reading));
+    magx /= 1 + _mag_comp.tcs.x * (temp - _mag_comp.t0_reading);
+
+    magy = ((1 + _mag_comp.sensit_coef.y) * magy) + _mag_comp.offset_coef.y + (_mag_comp.tco.y * (temp - _mag_comp.t0_reading));
+    magy /= 1 + _mag_comp.tcs.y * (temp - _mag_comp.t0_reading);
+
+    magz = ((1 + _mag_comp.sensit_coef.z) * magz) + _mag_comp.offset_coef.z + (_mag_comp.tco.z * (temp - _mag_comp.t0_reading));
+    magz /= 1 + _mag_comp.tcs.z * (temp - _mag_comp.t0_reading);
+
+    const float cr_ax_comp_x = (magx - _mag_comp.cross_axis.cross_x_y * magy) / (1 - _mag_comp.cross_axis.cross_y_x * _mag_comp.cross_axis.cross_x_y);
+    const float cr_ax_comp_y = (magy - _mag_comp.cross_axis.cross_y_x * magx) / (1 - _mag_comp.cross_axis.cross_y_x * _mag_comp.cross_axis.cross_x_y);
+    const float cr_ax_comp_z = (magz + (magx * (_mag_comp.cross_axis.cross_y_x * _mag_comp.cross_axis.cross_z_y - _mag_comp.cross_axis.cross_z_x) - 
+                                        magy * (_mag_comp.cross_axis.cross_z_y - _mag_comp.cross_axis.cross_x_y * _mag_comp.cross_axis.cross_z_x)) /
+                                       (1 - _mag_comp.cross_axis.cross_y_x * _mag_comp.cross_axis.cross_x_y));
+
+    // Store in field vector and convert uT to milligauss
+    Vector3f field { cr_ax_comp_x * 10.0f, cr_ax_comp_y * 10.0f, cr_ax_comp_z * 10.0f };
+    accumulate_sample(field, _compass_instance);
```

Comment:
```suggestion
    // Store in field vector and convert uT to milligauss
    Vector3f field { cr_ax_comp_x, cr_ax_comp_y, cr_ax_comp_z };
    field *= 10.0f;
    accumulate_sample(field, _compass_instance);
```

Reply:
I did consider suggesting this change but decided against in this instance.  For no particular reason - perhaps just for re-testing reasons.